index.rst

```diff
@@ -13,17 +13,542 @@
+The switch from ``pipetask`` to workspaces wouldn't actually *require* any major changes to the prompt processing framework's architecture, because each prompt processing worker has its own full SQL-backed butler, with the fan-out to workers and long-term persistence represented as transfers between data repositories.
+Workspaces would be created and committed wholly within individual workers,  requiring some changes to low-level tooling like :class:`~lsst.ctrl.mpexec.SeparablePipelineExecutor`.
+The existence of a full Python interface for workspace-based execution may actually remove the need for separate pipeline executor classes entirely.
+
+An alternative approach in which workspaces are more tightly integrated with prompt processing is both intriguing and more disruptive.
+In this model, prompt processing would have a custom :class:`WorkspaceButler` implementation that represents a client of a central-butler workspace existing on a worker node.
+This would give prompt-processing complete control over quantum graph generation and how it is interleaved with execution, allowing the graph to be built incrementally as new information arrives (first the ``nextVisit`` event, then a ``raw`` file artifact, and then perhaps a ``raw`` for a second stamp), with each processing step starting as soon as possible.
+There would be no SQL database on the workers at all, and the :class:`WorkspaceButler` implementation would implement its own :meth:`~lsst.daf.butler.LimitedButler.get` and :meth:`~lsst.daf.butler.LimitedButler.put` implementations for use by tasks.
+This would be a natural place to put in-memory caching of datasets, while delegating to a regular :class:`~lsst.daf.butler.Datastore` when actual file I/O is desired.
+
+This model would not work if a :attr:`~lsst.daf.butler.CollectionType.RUN` collection is the (only) unit for commits and quantum graph generation, since we would want to commit results from each worker to the central butler on a per-data ID basis.
+To address this, a workspace may be created with *sharding dimensions*, which define data IDs that may be committed independently.
+This restricts the tasks that may be run in such a workspace to those that have at least those dimensions, ensuring that all output datasets in the workspace can be divided into disjoint sets according to those data IDs.
+Committing a data ID from a sharded workspace moves all datasets and provenance with that data ID to the corresponding :attr:`~lsst.daf.butler.CollectionType.RUN` collection in the central repo, removing it from the workspace (note that file artifacts remain where they are unless this is an external workspace), allowing the :attr:`~lsst.daf.butler.CollectionType.RUN` collection and its workspace to coexist for a while.
```

Comment:
I think everything we'd need to refuse repeated commits will be present in registry tables and queryable through regular registry interfaces without additional work.  I'd leave it up to the workspace implementation whether to actually check or trust the caller, as in some contexts it may be more efficient to guard against this at a higher level.